scipy/sparse/csgraph/_shortest_path.pyx

```diff
@@ -649,294 +661,134 @@ def dijkstra(csgraph, directed=True, indices=None,
-                    if return_pred:
-                        pred[i, j_current] = v.index
+        next_val = v.first + csr_weights[j]
+        if next_val <= limit:
+            if dist_matrix[j_current] > next_val:
+                dist_matrix[j_current] = next_val
+                # The same vertex may be pushed multiple times to the queue, but
+                # anything with suboptimal distance is ignored when poped
+                heap.push(dist_index_pair_t(-next_val, j_current))
+                if return_pred:
+                    pred[j_current] = v.second
+                if return_source:
+                    sources[j_current] = sources[v.second]
 
-@cython.boundscheck(False)
-cdef int _dijkstra_directed(
-            const int[:] source_indices,
-            const double[:] csr_weights,
-            const int[:] csr_indices,
-            const int[:] csr_indptr,
-            double[:, :] dist_matrix,
-            int[:, :] pred,
-            DTYPE_t limit) except -1:
-    cdef:
-        unsigned int Nind = dist_matrix.shape[0]
-        unsigned int N = dist_matrix.shape[1]
-        unsigned int i, k, j_source
-        int return_pred = (pred.size > 0)
-        FibonacciHeap heap
-        FibonacciNode *v
-        FibonacciNode* nodes = <FibonacciNode*> malloc(N *
-                                                       sizeof(FibonacciNode))
-    if nodes == NULL:
-        raise MemoryError("Failed to allocate memory in _dijkstra_directed")
-
-    for i in range(Nind):
-        j_source = source_indices[i]
-
-        for k in range(N):
-            initialize_node(&nodes[k], k)
-
-        dist_matrix[i, j_source] = 0
-        heap.min_node = NULL
-        insert_node(&heap, &nodes[j_source])
-
-        while heap.min_node:
-            v = remove_min(&heap)
-            v.state = SCANNED
-
-            _dijkstra_scan_heap(&heap, v, nodes,
-                                csr_weights, csr_indices, csr_indptr,
-                                pred, return_pred, limit, i)
-
-            # v has now been scanned: add the distance to the results
-            dist_matrix[i, v.index] = v.val
-
-    free(nodes)
-    return 0
 
 @cython.boundscheck(False)
-cdef int _dijkstra_directed_multi(
+cdef int _dijkstra(
             const int[:] source_indices,
             const double[:] csr_weights,
             const int[:] csr_indices,
             const int[:] csr_indptr,
+            const double[:] csrT_weights,
+            const int[:] csrT_indices,
+            const int[:] csrT_indptr,
             double[:] dist_matrix,
             int[:] pred,
             int[:] sources,
             DTYPE_t limit) except -1:
     cdef:
+        unsigned int Nind = source_indices.shape[0]
         unsigned int N = dist_matrix.shape[0]
-
+        unsigned int i, j_source
         int return_pred = (pred.size > 0)
+        int return_sources = (sources.size > 0)
+        int directed = (csrT_weights.size == 0)
```

Comment:
Nitpick: 

* three lines are still interacting with the interpreter. As these are 1-d arrays, we can replace ``.size`` with ``.shape[0]`` which won't interact with the interpreter. 
* ``int`` -> ``bint``  this is a special Cython type to represent C boolean values.


```suggestion
        bint return_pred = (pred.shape[0] > 0)
        bint return_sources = (sources.shape[0] > 0)
        bint directed = (csrT_weights.shape[0] == 0)
```